commit 3a2af7cccbbaf2362db9053a946a6084e12bfa73
Author: Jinbum Park <jinb.park7@gmail.com>
Date:   Tue Jul 31 23:10:40 2018 +0900

    mac80211_hwsim: Fix possible Spectre-v1 for hwsim_world_regdom_custom
    
    User controls @idx which to be used as index of hwsim_world_regdom_custom.
    So, It can be exploited via Spectre-like attack. (speculative execution)
    
    This kind of attack leaks address of hwsim_world_regdom_custom,
    It leads an attacker to bypass security mechanism such as KASLR.
    
    So sanitize @idx before using it to prevent attack.
    
    I leveraged strategy [1] to find and exploit this gadget.
    
    [1] https://github.com/jinb-park/linux-exploit/tree/master/exploit-remaining-spectre-gadget/
    
    Signed-off-by: Jinbum Park <jinb.park7@gmail.com>
    [johannes: unwrap URL]
    Signed-off-by: Johannes Berg <johannes.berg@intel.com>

diff --git a/drivers/net/wireless/mac80211_hwsim.c b/drivers/net/wireless/mac80211_hwsim.c
index 7d0b460868f9..80e2c8595c7c 100644
--- a/drivers/net/wireless/mac80211_hwsim.c
+++ b/drivers/net/wireless/mac80211_hwsim.c
@@ -33,6 +33,7 @@
 #include <net/net_namespace.h>
 #include <net/netns/generic.h>
 #include <linux/rhashtable.h>
+#include <linux/nospec.h>
 #include "mac80211_hwsim.h"
 
 #define WARN_QUEUE 100
@@ -3229,6 +3230,9 @@ static int hwsim_new_radio_nl(struct sk_buff *msg, struct genl_info *info)
 			kfree(hwname);
 			return -EINVAL;
 		}
+
+		idx = array_index_nospec(idx,
+					 ARRAY_SIZE(hwsim_world_regdom_custom));
 		param.regd = hwsim_world_regdom_custom[idx];
 	}